ring_theory/ideals.lean

```diff
@@ -0,0 +1,49 @@
+import algebra.ring algebra.module data.set.basic
+
+universe u
+
+class is_prime_ideal {α : Type u} [comm_ring α] (S : set α) extends is_submodule S : Prop :=
+(ne_univ : S ≠ set.univ)
+(mem_or_mem_of_mul_mem : ∀ {x y : α}, x * y ∈ S → x ∈ S ∨ y ∈ S)
+
+theorem mem_or_mem_of_mul_eq_zero {α : Type u} [comm_ring α] (S : set α) [is_prime_ideal S] :
+  ∀ {x y : α}, x * y = 0 → x ∈ S ∨ y ∈ S :=
+λ x y hxy, have x * y ∈ S, by rw hxy; from (@is_submodule.zero α α _ _ S _ : (0:α) ∈ S),
+is_prime_ideal.mem_or_mem_of_mul_mem this
+
+class is_maximal_ideal {α : Type u} [comm_ring α] (S : set α) extends is_submodule S : Prop :=
+mk' ::
+  (ne_univ : S ≠ set.univ)
+  (eq_or_univ_of_subset : ∀ (T : set α) [is_submodule T], S ⊆ T → T = S ∨ T = set.univ)
+
+theorem is_maximal_ideal.mk {α : Type u} [comm_ring α] (S : set α) [is_submodule S] :
+  (1:α) ∉ S → (∀ x (T : set α) [is_submodule T], S ⊆ T → x ∉ S → x ∈ T → (1:α) ∈ T) → is_maximal_ideal S :=
+λ h₁ h₂,
+{ _inst_2 with
```

Comment:
Don't write `_inst_2`, give the type class instance a name: `[m : is_submodule S]` and use `m`.

Reply:
Also `{ ... with ...}` is outdated (it will vanish in the future). Use the following syntax `{ ne_univ := _, eq_or_univ_of_subset := _,  .. m }`